File: java/url-shortener/src/main/java/com/domain/urlshortener/core/codec/Base62.java
package com.domain.urlshortener.core.codec;

import java.math.BigInteger;

/**
 * @author: rocky.hu
 * @date: 2022/4/6 0:28
 */
public class Base62 {

    public static final BigInteger BASE = BigInteger.valueOf(62);
    public static final String DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
    public static final String REGEXP = "^[0-9A-Za-z]+$";

    public static String encode(BigInteger number) {
        if (number.compareTo(BigInteger.ZERO) == -1) {
            throw new IllegalArgumentException("number must not be negative");
        }
        StringBuilder result = new StringBuilder();
        while (number.compareTo(BigInteger.ZERO) == 1) {
            BigInteger[] divmod = number.divideAndRemainder(BASE);
            number = divmod[0];
            int digit = divmod[1].intValue();
            result.insert(0, DIGITS.charAt(digit));
        }
        return (result.length() == 0) ? DIGITS.substring(0, 1) : result.toString();
    }

    public static String encode(long number) {
        return encode(BigInteger.valueOf(number));
    }

    public static BigInteger decode(final String string) {
        if (string.length() == 0) {
            throw new IllegalArgumentException("string must not be empty");
        }
        BigInteger result = BigInteger.ZERO;
        int digits = string.length();
        for (int index = 0; index < digits; index++) {
            int digit = DIGITS.indexOf(string.charAt(digits - index - 1));
            result = result.add(BigInteger.valueOf(digit).multiply(BASE.pow(index)));
        }
        return result;
    }

}

File: backend/short-url/src/utils/base62.ts
class Base62 {
    private chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    private length = 8;

    constructor(chars?: string, length?: number) {
        this.chars = chars || this.chars;
        this.length = length || this.length;
    }

    public encode(num: number) {
        let result = '';
        while (num > 0) {
            result = this.chars.charAt(num % 62) + result;
            num = Math.floor(num / 62);
        }
        return result.padStart(this.length, this.chars[0]);
    }

    public decode(code: string): number {
        let data = 0;
        for (const c of code) {
            const i = this.chars.indexOf(c);
            data = data * this.chars.length + i;
        }
        return data;
    }
}

export default Base62;
